src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -262,8 +297,92 @@ public async ValueTask UnsubscribeAsync(int sid, int? maxMsgs, CancellationToken
     // only used for internal testing
     internal bool TestStallFlush() => _channelLock.Writer.TryWrite(1);
 
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowOnDisconnected() => throw new NatsException("Connection hasn't been established yet.");
+    private static async Task ReaderLoopAsync(ILogger<CommandWriter> logger, ISocketConnection connection, PipeReader pipeReader, Channel<int> channelSize, CancellationToken cancellationToken)
+    {
+        try
+        {
+            while (true)
+            {
+                var result = await pipeReader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (result.IsCanceled)
+                {
+                    break;
+                }
+
+                var buffer = result.Buffer;
+                var completed = buffer.Start;
```

Comment:
`consumed` may be a better name - since `pipeReader.AdvanceTo` takes 2 args - `consumed` and `examined`
need to track `examined` also

```
var examined = buffer.Start
```